[Metrics][Discover] Refactor chart layers hook

simianhacker opened the pull request:
## 🍒 Summary

This change refactors the `useChartLayers` hook in the Unified Metrics Grid to generate the layers based on the metric fields instead of the ESQL query. The previous `useChartLayers` hook created a request with `LIMIT 0` to get the columns for the chart to build the layers, this created an unnecessary network request for each chart. 

The change also introduces a ES|QL parameter substitution utility to ensure the columns defined in the chart layers matches the output from the ESQL query. It also introduces a utility to generate the metric aggregation and time bucket aggregation uses for both the chart layers and ESQL creation utility to ensure they match.

## 🛠️ Changes

- Created a comprehensive unit test suite for the new `useChartLayers` hook to ensure its correctness.
- Updated the `Chart` component and `useLensProps` hook to consume the new `useChartLayers` hook, simplifying their implementation.
- Added a new `create_aggregation.ts` utility with a `replaceFunctionParams` function to handle ES|QL parameter substitution and escaping.
- Included a unit test for the new aggregation helpers to verify their behavior.
- Added JSDoc documentation blocks to `createMetricAggregation` and `createTimeBucketAggregation` functions for better code clarity.

## 🎙️ Prompts

- "Can you create a function like `escapeString("system.load.1m")` that would return "system.load.`1m`" using the same technique."
- "can you add that to @src/platform/packages/shared/kbn-unified-metrics-grid/src/common/utils/esql/create_aggregation.ts"
- "Hum... Modify it to do this `replaceFunctionParams('AVG(??metricField)', { metricField: 'system.load.1m' })` which would output `AVG(system.load.\`1m\`)`"
- "I think you miss understood what I want. I want a function that I can call with `AVG(??metricField)` and `{ metricField: 'system.load.1m' }` and it will do the substitution using the AST."
- "Params and replaceParameters are not exported from the `esql-composer` library. Can you read @src/platform/packages/shared/kbn-esql-composer/README.md Could use this library to basically do the same thing?"
- "Can you write a unit test for this?"
- "I feel like what you originally wrote should work. It's working pretty much the same way here: @src/platform/packages/shared/kbn-unified-metrics-grid/src/common/utils/esql/create_esql_query.ts"
- "Can you add doc blocks to the other functions in the create_aggregation file?"
- "Can you convert the `const chartLayers = useMemo...` in @src/platform/packages/shared/kbn-unified-metrics-grid/src/components/chart/index.tsx to a hook called `useChartLayers` and put it in src/platform/packages/shared/kbn-unified-metrics-grid/src/components/chart/hooks/use_chart_layers.ts and add a corresponding test?"

🤖 This pull request was assisted by Gemini CLI


<!--ONMERGE {"backportTargets":["9.2"]} ONMERGE-->

simianhacker committed:
Refactor chart layers hook and add documentation

- Created a comprehensive unit test suite for the new `useChartLayers` hook to ensure its correctness.
- Updated the `Chart` component and `useLensProps` hook to consume the new `useChartLayers` hook, simplifying their implementation.
- Added a new `create_aggregation.ts` utility with a `replaceFunctionParams` function to handle ES|QL parameter substitution and escaping.
- Included a unit test for the new aggregation helpers to verify their behavior.
- Added JSDoc documentation blocks to `createMetricAggregation` and `createTimeBucketAggregation` functions for better code clarity.

Prompts:

- "Can you create a function like `escapeString("system.load.1m")` that would return `"system.load.`1m`"` using the same technique."
- "can you add that to @src/platform/packages/shared/kbn-unified-metrics-grid/src/common/utils/esql/create_aggregation.ts"
- "Hum... Modify it to do this `replaceFunctionParams('AVG(??metricField)', { metricField: 'system.load.1m' })` which would output `AVG(system.load.`1m`)`"
- "I think you miss understood what I want. I want a function that I can call with `AVG(??metricField)` and `{ metricField: 'system.load.1m' }` and it will do the substitution using the AST."
- "Params and replaceParameters are not exported from the `esql-composer` library. Can you read @src/platform/packages/shared/kbn-unified-metrics-grid/src/common/utils/esql/create_esql_query.ts Could use this library to basically do the same thing?"
- "Can you write a unit test for this?"
- "I feel like what you originally wrote should work. It's working pretty much the same way here: @src/platform/packages/shared/kbn-unified-metrics-grid/src/common/utils/esql/create_esql_query.ts"
- "Can you add doc blocks to the other functions in the create_aggregation file?"
- "Can you convert the `const chartLayers = useMemo...` in @src/platform/packages/shared/kbn-unified-metrics-grid/src/components/chart/index.tsx to a hook called `useChartLayers` and put it in src/platform/packages/shared/kbn-unified-metrics-grid/src/components/chart/hooks/use_chart_layers.ts and add a corresponding test?"
- "It looks like when you rewrote the chart/index.tsx you removed a bunch of imports"

🤖 This commit was assisted by Gemini CLI

```

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/346963)
* Commit: 9dd00221e51f9b7e74c9194309df0ad43d05ec02

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/346963#0199a744-8bc3-43a2-bc2e-b4c20cf333e8)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346963","number":346963,"commit":"9dd00221e51f9b7e74c9194309df0ad43d05ec02"}],"number":346963}
buildkite-pr-comment-->

simianhacker committed:
remvoing escapePlaceHolders

simianhacker committed:
Refactor ES|QL parameter replacement and add tests

- Refactored `replaceFunctionParams` to use an AST-based approach for more robust ES|QL parameter substitution.
- Removed an unnecessary test case for multiple `??` placeholders to align with primary use cases.
- Added a new test case for `SUM(RATE(??metricField))` to ensure proper handling of nested functions.
- Exported `replaceParameters` from `@kbn/esql-composer` to be used in the refactored function.

Prompts:

- "Can you purpose an alternative to the `replaceFunctionParams` in this file @/Users/ccowan/Projects/kibana/src/platform/packages/shared/kbn-unified-metrics-grid/src/common/utils/esql/create_aggregation.ts ? The second step where the function is being extracted using the `statsPrefix` feels a little hacky."
- "I update the function a little and added a console log so you can get a sense of what state the AST is in. Can you read the currrent file (don't remove my changes) then run the test?"
- "Can you add a test for `SUM(RATE(??metricField))`"

🤖 This commit was assisted by Gemini CLI

simianhacker committed:
Update chart hook tests and mock chartLayers

- Mocked the `chartLayers` attribute and added it to `useLensProps` hook calls in the test file.
- Removed an unnecessary assertion from the `useChartLayers` test file.

Prompts:

- "You review /Users/ccowan/Projects/kibana/src/platform/packages/shared/kbn-unified-metrics-grid/src/components/chart/hooks/use_lens_props.test.ts /Users/ccowan/Projects/kibana/src/platform/packages/shared/kbn-unified-metrics-grid/src/components/chart/hooks/use_lens_props.ts, mock the chartLayers attribute and add it to the hook calls?"

🤖 This commit was assisted by Gemini CLI

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/346973)
* Commit: e5f9de8dfdd37b2b2c3073f5f0f488a088d36d57

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/346973#0199a7e1-533c-41d1-9514-71582849d2ec)

### History
* :broken_heart: [Build #346963](https://buildkite.com/elastic/kibana-pull-request/builds/346963) failed 9dd00221e51f9b7e74c9194309df0ad43d05ec02

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346973","number":346973,"commit":"e5f9de8dfdd37b2b2c3073f5f0f488a088d36d57"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346963","number":346963,"commit":"9dd00221e51f9b7e74c9194309df0ad43d05ec02"}],"number":346973}
buildkite-pr-comment-->

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/refactor-chart-layers

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/347229)
* Commit: 74096768d74c39a6d76133febb671f8c89608b81

### Failed CI Steps


### History
* :broken_heart: [Build #346973](https://buildkite.com/elastic/kibana-pull-request/builds/346973) failed e5f9de8dfdd37b2b2c3073f5f0f488a088d36d57
* :broken_heart: [Build #346963](https://buildkite.com/elastic/kibana-pull-request/builds/346963) failed 9dd00221e51f9b7e74c9194309df0ad43d05ec02

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/347229","number":347229,"commit":"74096768d74c39a6d76133febb671f8c89608b81"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346973","number":346973,"commit":"e5f9de8dfdd37b2b2c3073f5f0f488a088d36d57"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346963","number":346963,"commit":"9dd00221e51f9b7e74c9194309df0ad43d05ec02"}],"number":347229}
buildkite-pr-comment-->

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/refactor-chart-layers

miloszmarcinkowski reviewed (approved):
LGTM, tested locally. I really like added JSDoc comments.

simianhacker committed:
removing extra isCommand

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/refactor-chart-layers

jennypavlova commented on src/platform/packages/shared/kbn-unified-metrics-grid/src/components/chart/index.tsx:
```diff
@@ -72,18 +73,17 @@ export const Chart = ({
   const lensProps = useLensProps({
     title: metric.name,
     query: esqlQuery,
-    unit: metric.unit,
```
Q: Why do we remove the unit? Is it part of `chartLayers`?

jennypavlova replied:
Thanks for explaining!

simianhacker committed:
Adding @kbn/test to tsconfig.json

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/347802)
* Commit: 5798bdf119a146789943e99c5baf5dace4e3e034

### Failed CI Steps
* [Quick Checks](https://buildkite.com/elastic/kibana-pull-request/builds/347802#0199ba04-c434-4e5a-91b7-e63c81ee9d9a)

### History
* :broken_heart: [Build #347760](https://buildkite.com/elastic/kibana-pull-request/builds/347760) failed 13697b3ccd7d7cb6eeb5f91395a5159ce6c9cfe9
* :broken_heart: [Build #347346](https://buildkite.com/elastic/kibana-pull-request/builds/347346) failed a08fa617d02dfe31847bf669ccdbd2f8e7c8b2a9
* :broken_heart: [Build #347229](https://buildkite.com/elastic/kibana-pull-request/builds/347229) failed 74096768d74c39a6d76133febb671f8c89608b81
* :broken_heart: [Build #346973](https://buildkite.com/elastic/kibana-pull-request/builds/346973) failed e5f9de8dfdd37b2b2c3073f5f0f488a088d36d57
* :broken_heart: [Build #346963](https://buildkite.com/elastic/kibana-pull-request/builds/346963) failed 9dd00221e51f9b7e74c9194309df0ad43d05ec02

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/347802","number":347802,"commit":"5798bdf119a146789943e99c5baf5dace4e3e034"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/347760","number":347760,"commit":"13697b3ccd7d7cb6eeb5f91395a5159ce6c9cfe9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/347346","number":347346,"commit":"a08fa617d02dfe31847bf669ccdbd2f8e7c8b2a9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/347229","number":347229,"commit":"74096768d74c39a6d76133febb671f8c89608b81"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346973","number":346973,"commit":"e5f9de8dfdd37b2b2c3073f5f0f488a088d36d57"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/346963","number":346963,"commit":"9dd00221e51f9b7e74c9194309df0ad43d05ec02"}],"number":347802}
buildkite-pr-comment-->

simianhacker committed:
Updating tsconfig to remove @kbn/esql-utils and @kbn/test

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/refactor-chart-layers

jennypavlova reviewed (approved):
The code LGTM 💯

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/refactor-chart-layers

simianhacker committed:
Merge branch 'main' of github.com:elastic/kibana into feature/refactor-chart-layers

kibanamachine commented:
Starting backport for target branches: 9.2

https://github.com/elastic/kibana/actions/runs/18322330349
